enh: add ENABLE_GMOCK option

cyrush opened the pull request:
Adds a option to enable google mock support (default = OFF)

cyrush commented:
To resolve #50.

We can discuss if the default should be `OFF` or `ON`, I choose `OFF` b/c of #55

cyrush commented:
For symmetry, we talked about also adding `ENABLE_GTEST` and `ENABLE_FRUIT`.
If possible, I would like to do this as part of another PR.

cyrush commented on cmake/BLTOptions.cmake:
```diff
@@ -61,6 +61,7 @@ option(ENABLE_MPI          "Enables MPI support" OFF)
 option(ENABLE_OPENMP       "Enables OpenMP compiler support" OFF)
 option(ENABLE_CUDA         "Enable CUDA support" OFF)
 
+option(ENABLE_GMOCK        "Enable Google Mock testing support" OFF)
```
It mirrors `ENABLE_MPI`, `ENABLE_OPENMP`, and `ENABLE_CUDA`.
It's also grouped with them in `BLTOptions.cmake`

I thought about adding a more descriptive comment that acknowledges the fact that this also screened by `ENABLE_TESTS` -- would that suffice?

kennyweiss replied:
I think so.

It would be good to think about a clean way to set up sub-options for a given TPL/component (e.g. to include the name of the TPL/component that is is related to).

kennyweiss replied:
E.g. it we had 
* ``TESTS_ENABLE_GMOCK`` 
* ``TESTS_ENABLE_GTEST``
* ``TESTS_ENABLE_FRUIT`` 
* ...

Then these would show up in the ``TESTS`` group in ``ccmake``/``cmake-gui`` rather than the ``ENABLE`` group.  
Pro: This would declutter the ``ENABLE`` group
Con: But, it might make the options less discoverable.

cyrush replied:
`TESTS_ENABLE_GMOCK ` to me implies "enable the gmock tests" vs build gmock

kennyweiss replied:
Yeah, it is definitely not perfect, and I don't think this should hold up the current PR.

But I was aiming for something clean along these lines.

davidbeckingsale commented on thirdparty_builtin/googletest-release-1.8.0/CMakeLists.txt:
```diff
@@ -16,13 +16,21 @@ project( googletest-distribution )
 
-option(BUILD_GTEST "Builds the googletest subproject" OFF)
+
+##########################################################
```
Yeah, I think it makes sense to do this at a higher level.

cyrush replied:
I'll look into it -- Note: We already have a BLT specific fix in this file on master, we should try to move that as well.

cyrush commented:
I moved our `ENABLE_GMOCK` logic into one of our `CMakeLists.txt` files, and added more details in `BLTOptions.cmake` related to the fact that `ENABLE_GMOCK` is also guarded by `ENABLE_TESTS`.

I also tried to move the policy command related to `project(zzz VERSION "x.y.z)` out of the master googletest `CMakeLists.txt` file, but that change has to stay there  to take effect -- I believe b/c of the use of the `cmake_minimum_version` command.

cyrush commented:
@kennyweiss are these changes sufficient to merge?

white238 commented on thirdparty_builtin/CMakeLists.txt:
```diff
@@ -57,18 +57,36 @@ if(ENABLE_TESTS)
+    # gtest's cmake project
+    #
+    
+    if(ENABLE_GMOCK)
```
This off/on and on/off logic doesn't make sense to me. Can you explain? Why does enabling gmock disable building gtest

cyrush replied:
It's the way that googletest's cmake project works.

First, I don't belive you can build gmock w/o gtest. 

You either build gmock (using their `BUILD_GMOCK=ON` which also builds gtest), or you build gtest (and don't build gmock)

I haven't checked, so I don't know if you can actually set both to OFF (for a future case where we would also want to support our own`ENABLE_GTEST` -- but we could always simply not include the projects as our own guard.

cyrush replied:
Check out how they are used in:

https://github.com/LLNL/blt/blob/master/thirdparty_builtin/googletest-release-1.8.0/CMakeLists.txt

While keeping in mind that building gmock also takes care of building gtest.
